11-designing-for-mental-health/projects/final/app/src/main/java/com/raywenderlich/android/tacotuesday/shared/CustomRatingBar.kt

```diff
@@ -0,0 +1,212 @@
+    return (event?.let { exploreByTouchHelper.dispatchHoverEvent(it) } ?: run { false }
+        || super.dispatchHoverEvent(event))
+  }
+
+  override fun dispatchKeyEvent(event: KeyEvent?): Boolean {
+    return (event?.let { exploreByTouchHelper.dispatchKeyEvent(it) } ?: run { false }
+        || super.dispatchKeyEvent(event))
+  }
+
+  override fun onFocusChanged(gainFocus: Boolean, direction: Int, previouslyFocusedRect: Rect?) {
+    super.onFocusChanged(gainFocus, direction, previouslyFocusedRect)
+    exploreByTouchHelper.onFocusChanged(gainFocus, direction, previouslyFocusedRect)
+  }
+
+  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+    super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+    val width: Int = maxRating * (diameter + spacingWidth) - spacingWidth + paddingLeft + paddingRight
+    val height: Int = diameter + paddingTop + paddingBottom
+
+    setMeasuredDimension(width, height)
+  }
+
+  override fun onDraw(canvas: Canvas) {
+    diameter / 2.toFloat()
+    for (i in 0 until maxRating) {
+      val image = if (i >= rating) {
+        ResourcesCompat.getDrawable(resources, drawableResUnchecked, context.theme)
+      } else {
+        ResourcesCompat.getDrawable(resources, drawableResChecked, context.theme)
+      }
+      image!!.setBounds(rectangles[i].left, rectangles[i].top, rectangles[i].right,
+          rectangles[i].bottom)
+      image.draw(canvas)
+    }
+  }
+
+  override fun onTouchEvent(event: MotionEvent): Boolean {
+    when (event.action) {
+      MotionEvent.ACTION_DOWN -> return true
+      MotionEvent.ACTION_UP -> {
+        val selectedIndex = findRatingAtPoint(event.x, event.y)
+        onSelected(selectedIndex)
+        return true
+      }
+    }
+    return super.onTouchEvent(event)
+  }
+
+  protected fun onSelected(selectedIndex: Int) {
+    if (selectedIndex == INVALID_VALUE) return
+    rating = selectedIndex + 1
+    invalidate()
+  }
+
+  private fun findRatingAtPoint(x: Float, y: Float): Int {
+    var selectedIndex = INVALID_VALUE
+    for (i in 0 until maxRating) {
+      if (rectangles[i].contains(x.toInt(), y.toInt())) {
+        selectedIndex = i
+        break
+      }
+    }
+    return selectedIndex
+  }
+
+  companion object {
+    const val INVALID_VALUE = -1
+    const val spacingWidthDp = 10f
+    const val iconSizeDp = 48f
+  }
+
+  inner class RatingExploreByTouchHelper(host: View) : ExploreByTouchHelper(host) {
+
+    override fun getVirtualViewAt(x: Float, y: Float): Int {
+      val index = findRatingAtPoint(x, y)
+      return if (index == INVALID_VALUE) INVALID_ID else index
+    }
+
+    override fun getVisibleVirtualViews(virtualViewIds: MutableList<Int>?) {
+      for (i in 0 until maxRating) virtualViewIds?.add(i)
+    }
+
+    override fun onPerformActionForVirtualView(
+        virtualViewId: Int,
+        action: Int,
+        arguments: Bundle?
+    ): Boolean {
+      when (action) {
+        AccessibilityNodeInfoCompat.ACTION_CLICK -> {
```

Comment:
I think you can use if statement here because you have only one action
... : Boolean = if (action == AccessibilityNodeInfoCompat.ACTION_CLICK) {
onSelected(virtualViewId)
true
} else {
false
}